learning.py

```diff
@@ -810,7 +815,110 @@ def init_examples(examples, idx_i, idx_t, o_units):
 
 
 def find_max_node(nodes):
-    return nodes.index(argmax(nodes, key=lambda node: node.value))
+    return nodes.index(max(nodes, key=lambda node: node.value))
+
+
+class BinarySVM:
+    def __init__(self, kernel=linear_kernel, C=1.0):
+        self.kernel = kernel
+        self.C = C
+        self.eps = 1e-6
+        self.n_sv = -1
+        self.sv_x, self.sv_y, = np.zeros(0), np.zeros(0)
+        self.alphas = np.zeros(0)
+        self.w = None
+        self.b = 0.0  # intercept
+
+    def fit(self, X, y):
+        # In QP formulation (dual): m variables, 2m+1 constraints (1 equation, 2m inequations)
+        self.QP(X, y)
+        sv_indices = list(filter(lambda i: self.alphas[i] > self.eps, range(len(y))))
+        self.sv_x, self.sv_y, self.alphas = X[sv_indices], y[sv_indices], self.alphas[sv_indices]
+        self.n_sv = len(sv_indices)
+        if self.kernel == linear_kernel:
+            self.w = np.dot(self.alphas * self.sv_y, self.sv_x)
+        # calculate b: average over all support vectors
+        sv_boundary = self.alphas < self.C - self.eps
+        self.b = np.mean(self.sv_y[sv_boundary] - np.dot(self.alphas * self.sv_y,
+                                                         self.kernel(self.sv_x, self.sv_x[sv_boundary])))
+
+    def QP(self, X, y):
+        # in QP formulation (dual): m variables, 2m+1 constraints (1 equation, 2m inequations)
+        m = len(y)  # m = n_samples
+        K = self.kernel(X)  # gram matrix
+        P = K * np.outer(y, y)
+        q = -np.ones(m)
+        G = np.vstack((-np.identity(m), np.identity(m)))
+        h = np.hstack((np.zeros(m), np.ones(m) * self.C))
+        A = y.reshape((1, -1))
+        b = np.zeros(1)
+        # make sure P is positive definite
+        P += np.eye(P.shape[0]).__mul__(1e-3)
+        self.alphas = solve_qp(P, q, G, h, A, b, sym_proj=True)
+
+    def project(self, x):
+        if self.w is None:
+            return np.dot(self.alphas * self.sv_y, self.kernel(self.sv_x, x)) + self.b
+        return np.dot(x, self.w) + self.b
+
+    def predict(self, x):
+        return np.sign(self.project(x))
+
+
+class MultiSVM:
+    def __init__(self, kernel=linear_kernel, decision_function='ovr', C=1.0):
+        self.kernel = kernel
+        self.decision_function = decision_function
+        self.C = C
+        self.n_class, self.classifiers = 0, []
+
+    def fit(self, X, y):
+        labels = np.unique(y)
+        self.n_class = len(labels)
+        if self.decision_function == 'ovr':  # one-vs-rest method
+            for label in labels:
```

Comment:
Again here, the variable names could be a bit more descriptive.

Reply:
Same here.